libraries/chain/market_object.cpp

```diff
@@ -0,0 +1,267 @@
+   i256 fp_coll_amt = feed_price.base.amount.value;
+
+   // firstly we calculate without the fraction (x), the result could be a bit too small
+   i256 numerator = fp_coll_amt * mp_debt_amt * debt.value * tcr
+                  - fp_debt_amt * mp_debt_amt * collateral.value * GRAPHENE_COLLATERAL_RATIO_DENOM;
+   if( numerator < 0 ) // feed protected, actually should not be true here, just check to be safe
+      return make_pair( asset( 0, collateral_type() ), asset( 0, debt_type() ) );
+
+   i256 denominator = fp_coll_amt * mp_debt_amt * tcr - fp_debt_amt * mp_coll_amt * GRAPHENE_COLLATERAL_RATIO_DENOM;
+   if( denominator <= 0 ) // black swan
+      return make_pair( get_collateral(), get_debt() );
+
+   // note: if add 1 here, will result in 1.5x imperfection rate;
+   //       however, due to rounding, the result could still be a bit too big, thus imperfect.
+   i256 to_cover_i256 = ( numerator / denominator );
+   if( to_cover_i256 >= debt.value ) // avoid possible overflow
+      return make_pair( get_collateral(), get_debt() );
+   share_type to_cover_amt = static_cast< int64_t >( to_cover_i256 );
+
+   // stabilize
+   // note: rounding up-down results in 3x imperfection rate in comparison to down-down-up
+   asset to_pay = asset( to_cover_amt, debt_type() ) * match_price;
+   asset to_cover = to_pay * match_price;
+   to_pay = to_cover.multiply_and_round_up( match_price );
+
+   if( to_cover.amount >= debt || to_pay.amount >= collateral ) // to be safe
+      return make_pair( get_collateral(), get_debt() );
+   FC_ASSERT( to_pay.amount < collateral && to_cover.amount < debt );
+
+   // check collateral ratio after filled, if it's OK, we return
+   price fp = asset( fp_coll_amt, collateral_type() ) / asset( fp_debt_amt, debt_type() );
+   price new_call_price = price::call_price( get_debt() - to_cover, get_collateral() - to_pay, tcr );
+   if( new_call_price > fp )
+      return make_pair( std::move(to_pay), std::move(to_cover) );
+
+   // be here, to_cover is too small due to rounding. deal with the fraction
+   numerator += fp_coll_amt * mp_debt_amt * tcr; // plus the fraction
+   to_cover_i256 = ( numerator / denominator ) + 1;
+   if( to_cover_i256 >= debt.value ) // avoid possible overflow
+      to_cover_i256 = debt.value;
+   to_cover_amt = static_cast< int64_t >( to_cover_i256 );
+
+   asset max_to_pay = ( ( to_cover_amt == debt.value ) ? get_collateral()
+                        : asset( to_cover_amt, debt_type() ).multiply_and_round_up( match_price ) );
+   if( max_to_pay.amount > collateral )
+      max_to_pay.amount = collateral;
+
+   asset max_to_cover = ( ( max_to_pay.amount == collateral ) ? get_debt() : ( max_to_pay * match_price ) );
+   if( max_to_cover.amount >= debt ) // to be safe
+   {
+      max_to_pay.amount = collateral;
+      max_to_cover.amount = debt;
+   }
+
+   if( max_to_pay <= to_pay || max_to_cover <= to_cover ) // strange data. should skip binary search and go on, but doesn't help much
```

Comment:
Actually since there is a `-` (minus sign) in denominator, typical rounding error here is not small. That's also why a binary search significantly improves perfection ratio (perfect means selling a small amount less will result in `CR <= max(TCR, MCR)`).

However, it's strange as I said "doesn't help much" here: theoretically it's better to search from here up to full debt, but it doesn't make significant difference on perfection ratio with my test case. Perhaps need special test cases for this scenario.

Reply:
Believe me, without the binary search, the result is much uglier:
* without binary search: 571702,66476,715,4010,8365,1381
* with binary search: 646007,1262,10,3,96,27

The arrays above are count of `perfect`, within `0.01%`, `0.1%`, `1%`, `10%`, others. They're not always small Satoshis, also sometimes a couple of Satoshis means a lot.

For example, the `8365` means there is a 1.3% chance (8k/~600k) that you're selling 10% more than minimum required.